clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -0,0 +1,229 @@
+            public void write(ByteBuffer buffer, Object o) {
+                Assignment assignment = (Assignment) o;
+                assignment.struct.writeTo(buffer);
+            }
+
+            @Override
+            public Object read(ByteBuffer buffer) {
+                Struct struct = (Struct) ASSIGNMENT_V0.read(buffer);
+                return new Assignment(struct);
+            }
+
+            @Override
+            public int sizeOf(Object o) {
+                Assignment assignment = (Assignment) o;
+                return assignment.struct.sizeOf();
+            }
+        };
+    }
+
+    public static class Assignment {
+        private final Struct struct;
+        private final short version;
+        private final List<TopicPartition> partitions;
+
+        private Map<String, List<Integer>> asMap(Collection<TopicPartition> partitions) {
+            Map<String, List<Integer>> partitionMap = new HashMap<>();
+            for (TopicPartition partition : partitions) {
+                String topic = partition.topic();
+                List<Integer> topicPartitions = partitionMap.get(topic);
+                if (topicPartitions == null) {
+                    topicPartitions = new ArrayList<>();
+                    partitionMap.put(topic, topicPartitions);
+                }
+                topicPartitions.add(partition.partition());
+            }
+            return partitionMap;
+        }
+
+        public Assignment(List<TopicPartition> partitions) {
+            this.struct = new Struct(ASSIGNMENT_V0);
+
+            List<Struct> topicAssignments = new ArrayList<>();
+            for (Map.Entry<String, List<Integer>> topicEntry : asMap(partitions).entrySet()) {
+                Struct topicAssignment = new Struct(TOPIC_ASSIGNMENT_V0);
+                topicAssignment.set(TOPIC_KEY_NAME, topicEntry.getKey());
+                topicAssignment.set(PARTITIONS_KEY_NAME, topicEntry.getValue().toArray());
+                topicAssignments.add(topicAssignment);
+            }
+            this.struct.set(ASSIGNMENT_KEY_NAME, topicAssignments.toArray());
+            this.partitions = partitions;
+
+            this.version = CONSUMER_PROTOCOL_V0;
+            this.struct.set(VERSION_KEY_NAME, CONSUMER_PROTOCOL_V0);
+        }
+
+        public Assignment(Struct struct) {
+            this.struct = struct;
+            this.partitions = new ArrayList<>();
+            for (Object structObj : struct.getArray(ASSIGNMENT_KEY_NAME)) {
+                Struct assignment = (Struct) structObj;
+                String topic = assignment.getString(TOPIC_KEY_NAME);
+                for (Object partitionObj : assignment.getArray(PARTITIONS_KEY_NAME)) {
+                    Integer partition = (Integer) partitionObj;
+                    this.partitions.add(new TopicPartition(topic, partition));
+                }
+            }
+            this.version = struct.getShort(VERSION_KEY_NAME);
+        }
+
+        public short version() {
```

Comment:
@ewencp @guozhangwang I've added a commit to clarify current versioning assumptions of the embedded metadata/assignment format. Let me know if that is sufficient for now. I think we'll still want to consider this in a little more detail in a follow-up issue after this is checked in.